subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -1023,7 +1023,13 @@ public ImmutableAttributesFactory getAttributesFactory() {
      * There may be some incoming edges left at that point, but they must all be coming from constraints.
      */
     public void clearConstraintEdges(PendingDependencies pendingDependencies, NodeState backToPendingSource) {
-        for (EdgeState incomingEdge : incomingEdges) {
+        if (incomingEdges.isEmpty()) {
+            return;
+        }
+        // Cleaning has to be done on a copied collection because of the recompute happening on selector removal
+        List<EdgeState> remainingIncomingEdges = ImmutableList.copyOf(incomingEdges);
```

Comment:
I am surprised this has no perf impact, this is likely to be the cause, if any.

Reply:
For that one we most likely not have a perf test with tons of constraints. However the list will be empty or have one element in most cases. To have any meaningful size here, you need multiple constraints on the same module.